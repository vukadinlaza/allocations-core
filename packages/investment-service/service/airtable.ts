/* eslint-disable @typescript-eslint/no-non-null-assertion */
import fetch from "node-fetch";
import { Investment } from "@allocations/core-models";
import { NameScanData } from "./utils/namescan";
const { AIRTABLE_API_KEY, AIRTABLE_DEAL_TRACKER_BASE_ID } = process.env;

const AIRTABLE_DEALS_TABLE_ID = "Deals";
const AIRTABLE_USERS_TABLE_ID = "Users";
const AIRTABLE_DEAL_TRACKER_TABLE_ID =
  process.env.AIRTABLE_DEAL_TRACKER_TABLE_ID || "Deal Tracker STAGING";

if (!AIRTABLE_DEAL_TRACKER_BASE_ID || !AIRTABLE_DEAL_TRACKER_BASE_ID) {
  throw new Error(
    `Environment variables AIRTABLE_API_KEY and  AIRTABLE_DEAL_TRACKER_BASE_ID are required`
  );
}

type RecordId = string;

type DealTrackerInvestment = {
  record_id: RecordId;
  "Investor Name"?: string;
  Email?: string[];
  "Total Amount Committed (without fees)"?: number;
  "Total Amount Committed (with fees)"?: number;
  "Deal Name (webapp)"?: string[];
  "Investor Name/Entity"?: string;
  "Investor Type"?: string;
  "Investor State"?: string;
  "Wire Name"?: string[];
  "Investor Country"?: string;
  "Accredited Investor"?: string;
  "Capital Call"?: boolean;
  "Upcoming Capital Call Date"?: string;
  "SPV/Fund Name Lookup"?: string[];
  "Upcoming Capital Call (%)"?: number;
  "*Bank Account Name (from Wire Feed)"?: string[];
  "*Date (from Wire Feed)"?: string[];
  "*Investor Name (from Wire Feed)"?: string[];
  "Type (from Bookings by month)"?: string[]; // Fund, Entity, Individual
  "Portfolio Wire Date (from Bookings by month)"?: string[];
  "SPV/Fund Name (from Bookings by month)"?: string[];
  "*Contribution 1 (from Wire Feed)"?: string[];
  "Net Amount Contributed"?: number | string;
  "Aggregate Contributed (%)"?: number;
  "Gross Contribution Outstanding (%)"?: number | string;
  "Upcoming Capital Call Amount"?: number;
  "Deal ID"?: string[];
  "Mongo Investment ID"?: string;
  "Blue Sky Fees"?: number;
  createdTime?: string;
  "KYC Complete"?: string;
  "KYC Result"?: string[];
  "KYC Scan Meta Data"?: string;
};

const getNameOrEntity = (investment: Investment) => {
  return investment.investor_type === "Entity"
    ? investment.investor_entity_name
    : investment.investor_name;
};

const findDealRecordId = async (deal_id: string): Promise<RecordId> => {
  const filterByFormula = encodeURIComponent(`FIND("${deal_id}", {deal_id})`);
  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_DEALS_TABLE_ID}?filterByFormula=${filterByFormula}`,
    {
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
      },
    }
  );

  if (!res.ok) throw new Error("Failed to fetch from Deals Airtable");

  const { records } = await res.json();
  if (!records.length)
    throw new Error(`Unable to find deal with id ${deal_id}`);

  return records[0].id;
};

const findUserRecordId = async (user_id: string): Promise<RecordId> => {
  const filterByFormula = encodeURIComponent(`FIND("${user_id}", {_id})`);
  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_USERS_TABLE_ID}?filterByFormula=${filterByFormula}`,
    {
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
      },
    }
  );

  if (!res.ok) throw new Error("Failed to fetch from Users Airtable");

  const { records } = await res.json();
  if (!records.length)
    throw new Error(`Unable to find user with id ${user_id}`);

  return records[0].id;
};

export const createInvestment = async (
  investment: Investment
): Promise<DealTrackerInvestment> => {
  const deal_record_id = await findDealRecordId(
    investment.metadata.get("deal_id")
  );

  const user_record_id = await findUserRecordId(investment.user_id.toString());

  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_DEAL_TRACKER_TABLE_ID}`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        records: [
          {
            fields: {
              "Investor Name": investment.investor_name,
              "Deal Name (webapp)": [deal_record_id],
              Email: [user_record_id],
            },
          },
        ],
      }),
    }
  );

  if (!res.ok) {
    throw new Error("Failed to create deal tracker investment on Airtable");
  }

  const { records } = await res.json();
  return { record_id: records[0].id, ...records[0].fields };
};

export const updateSignedInvestment = async (
  investment: Investment
): Promise<DealTrackerInvestment> => {
  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_DEAL_TRACKER_TABLE_ID}`,
    {
      method: "PATCH",
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        records: [
          {
            id: investment.metadata.get("deal_tracker_record_id"),
            fields: {
              "Carry %": investment.carry_fee_percent,
              "Management Fee %": investment.management_fee_percent,
              "Investor Type": investment.investor_type,
              "Investor Name/Entity": getNameOrEntity(investment),
              "Investor Country": investment.investor_country,
              "Accredited Investor": investment.accredited_investor_type,
              "Investor State": investment.investor_state,
              "Total Amount Committed (without fees)":
                investment.total_committed_amount,
              "Mongo Investment ID": investment._id,
            },
          },
        ],
      }),
    }
  );
  if (!res.ok) throw new Error("Failed to update Airtable");

  const { records } = await res.json();
  return { record_id: records[0].id, ...records[0].fields };
};

export const findDealTrackerInvestment = async (
  investment: Investment
): Promise<DealTrackerInvestment | null> => {
  const deal_id = investment.metadata.get("deal_id");

  const name =
    investment.investor_type === "Entity" ||
    investment.investor_type === "Trust"
      ? investment.investor_entity_name
      : investment.investor_name;

  const nameOrEntity = getNameOrEntity(investment);
  const filterByFormula = encodeURIComponent(
    `AND(FIND("${investment.user_id}", {Mongo User ID}), FIND("${deal_id}", {deal_id}), FIND("${name}", {Investor Name/Entity}))`
  );
  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_DEAL_TRACKER_TABLE_ID}?filterByFormula=${filterByFormula}`,
    {
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
      },
    }
  );

  if (!res.ok) throw new Error("Failed to fetch from Airtable");

  const { records } = (await res.json()) as any;
  const record = records.reduce((foundRecord: any, record: any) => {
    if (record["Investor Name/Entity"] === nameOrEntity) return record;
    else if (!foundRecord && !record["Investor Name/Entity"]) return record;
    else return record;
  }, null);
  if (!record) return null;

  return { record_id: record.id, ...record.fields };
};

export const createOrFindDealTrackerInvestment = async (
  investment: Investment
): Promise<DealTrackerInvestment> => {
  const dealTrackerInvestment = await findDealTrackerInvestment(investment);
  if (dealTrackerInvestment) return dealTrackerInvestment;

  return createInvestment(investment);
};

export const updateBlueSkyFees = async (
  investment: Investment,
  newFeeValue: string
): Promise<DealTrackerInvestment> => {
  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_DEAL_TRACKER_TABLE_ID}`,
    {
      method: "PATCH",
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        records: [
          {
            // @ts-ignore
            id: investment.id,
            fields: {
              "Blue Sky Fees": newFeeValue,
            },
          },
        ],
      }),
    }
  );

  if (!res.ok)
    throw new Error("Failed to update Airtable with Blue Sky fee state value");

  const { records } = await res.json();
  return { record_id: records[0].id, ...records[0].fields };
};

export const getDealTrackerDealRecords = async (
  deal_id: string
): Promise<DealTrackerInvestment | null> => {
  const filterByFormula = encodeURIComponent(
    `AND(FIND("${deal_id}", {deal_id}))`
  );
  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_DEAL_TRACKER_TABLE_ID}?filterByFormula=${filterByFormula}`,
    {
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
      },
    }
  );

  if (!res.ok) throw new Error("Failed to fetch from Airtable");

  const { records } = (await res.json()) as any;

  return records.map((r: any) => {
    return {
      id: r.id,
      ...r.fields,
    };
  });
};

export const updateKYCData = async (
  investment: Investment,
  nameScanData: NameScanData
): Promise<DealTrackerInvestment> => {
  const res = await fetch(
    `https://api.airtable.com/v0/${AIRTABLE_DEAL_TRACKER_BASE_ID}/${AIRTABLE_DEAL_TRACKER_TABLE_ID}`,
    {
      method: "PATCH",
      headers: {
        Authorization: `Bearer ${AIRTABLE_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        records: [
          {
            id: investment.metadata.get("deal_tracker_record_id"),
            fields: {
              "KYC Complete": nameScanData?.scanned,
              "KYC Result": nameScanData?.singleHitType || "N/A",
              "KYC Scan Meta Data": nameScanData?.metaData,
            },
          },
        ],
      }),
    }
  );

  if (!res.ok) throw new Error("Failed to update Airtable KYC Data");

  const { records } = await res.json();
  return { record_id: records[0].id, ...records[0].fields };
};
