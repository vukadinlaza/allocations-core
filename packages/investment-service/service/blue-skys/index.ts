import { Investment } from "@allocations/core-models";
import { connectMongoose } from "@allocations/service-common";
import { updateAirtableBSF, getNewBSFees } from "../utils/blueSkyFees";

export const handler = async ({ Records }: any): Promise<void> => {
  await connectMongoose();
  for (const record of Records) {
    const { _id, metadata } = JSON.parse(record.Sns.Message);
    const investment = await Investment.findById(_id);

    try {
      if (!investment) {
        throw new Error(`Unable to find investment with id ${_id}`);
      }
      const investmentsWithBSF = await getNewBSFees(metadata.deal_id);
      await updateAirtableBSF(investmentsWithBSF, metadata.deal_id);
    } catch (err) {
      console.error("Error syncing blue sky fees to deal tracker", err);
    }
  }
};
