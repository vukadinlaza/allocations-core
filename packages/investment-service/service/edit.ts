import { Investment, InvestmentAgreement } from "@allocations/core-models";
import {
  connectMongoose,
  parseRequest,
  send,
  sendError,
  LambdaEvent,
  triggerTransition,
  LambdaResponse,
} from "@allocations/service-common";

export const handler = async (event: LambdaEvent): Promise<LambdaResponse> => {
  try {
    await connectMongoose();
    const { body } = parseRequest(event);
    const investment = await Investment.findOne({
      _id: body.id,
      phase: ["signed", "agreements-pending"],
    });

    if (!investment) {
      return sendError({
        status: "404",
        error: new Error(
          "Investment in phase signed or agreements-pending not found"
        ),
      });
    }

    await InvestmentAgreement.updateMany(
      { investment_id: investment._id, status: { $ne: "archived" } },
      { status: "archived" }
    );

    await triggerTransition({
      id: body.id,
      action: "EDIT",
      phase: investment.phase,
    });

    return send({ acknowledged: true });
  } catch (err: any) {
    console.error(err);
    return sendError({ error: err as Error, status: "500" });
  }
};
