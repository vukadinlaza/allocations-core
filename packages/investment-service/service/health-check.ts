import { send } from "@allocations/service-common";

export const HEALTH_CHECK_PATH = "health-check";

module.exports.handler = async () => {
  return send({ message: "Healthy!" });
};
