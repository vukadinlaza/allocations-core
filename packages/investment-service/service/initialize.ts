import { Investment, Document } from "@allocations/core-models";
import {
  connectMongoose,
  parseRequest,
  send,
  sendError,
  LambdaEvent,
  triggerTransition,
  LambdaResponse,
} from "@allocations/service-common";
import fetch from "node-fetch";

export const handler = async (event: LambdaEvent): Promise<LambdaResponse> => {
  try {
    await connectMongoose();
    const { body } = parseRequest(event);
    const investment = await Investment.findOne({
      _id: body.id,
      phase: "new",
    });
    console.log({ investment, body });

    if (!investment) {
      return send({ acknowledged: true });
    }

    const dealRes = await fetch(
      `${process.env.BUILD_API_URL}/api/v1/deals/${investment.metadata.get(
        "deal_id"
      )}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
          "X-API-TOKEN": event.headers["X-API-TOKEN"],
        },
      }
    );
    console.log({ dealRes });
    const deal = await dealRes.json();

    const document = await Document.create({
      investment_id: body.id,
      title: `Private Fund Documents - ${deal.name}`,
      bucket: "",
      path: "",
      content_type: "",
      complete: false,
      created_by: body.investor_email,
    });

    console.log({ document });

    if (investment.metadata.get("invited")) {
      await triggerTransition({
        id: investment._id.toString(),
        action: "CREATED_WITH_INVITE",
        phase: "new",
      });

      return send({ acknowledged: true });
    }
    console.log("ROPIV", process.env.SERVICE_TOPIC);
    await triggerTransition({
      id: investment._id.toString(),
      action: "CREATED",
      phase: "new",
    });

    return send({ acknowledged: true });
  } catch (err: any) {
    console.error(err);
    return sendError({ error: err as Error, status: "500" });
  }
};
