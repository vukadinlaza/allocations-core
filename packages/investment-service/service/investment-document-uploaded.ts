import { connectMongoose } from "@allocations/service-common";
import { S3Event } from "aws-lambda";
import { Document } from "@allocations/core-models";

export const handler = async ({ Records }: S3Event) => {
  await connectMongoose();
  for (const record of Records) {
    console.log({ record });
    console.log(JSON.stringify(record, null, 2));
    const document = await Document.findOneAndUpdate(
      {
        investment_id: record.s3.object.key.split("/")[1],
      },
      {
        complete: true,
        bucket: record.s3.bucket.name,
        path: record.s3.object.key,
        content_type: record.s3.object.key.split(".")[1],
      }
    );

    if (!document) {
      throw new Error(`Unable to find document: ${record.s3.object.key}`);
    }
  }
};
