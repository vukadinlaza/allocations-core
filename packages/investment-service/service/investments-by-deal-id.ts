import { Investment } from "@allocations/core-models";
import { Types } from "mongoose";
import {
  connectMongoose,
  parseRequest,
  send,
  sendError,
  LambdaEvent,
  LambdaResponse,
} from "@allocations/service-common";

export const handler = async (event: LambdaEvent): Promise<LambdaResponse> => {
  try {
    await connectMongoose();
    const { params } = parseRequest(event);
    const { deal_id } = params;

    if (!deal_id || !Types.ObjectId.isValid(deal_id)) {
      return sendError({
        error: new Error(`invalid or missing deal_id: ${deal_id}`),
        status: "400",
      });
    }

    const investmentsByDealId = await Investment.find({
      "metadata.deal_id": deal_id,
    });
    return send({ investments: investmentsByDealId });
  } catch (err: any) {
    console.error(err);
    return sendError({ error: err as Error, status: "500" });
  }
};
