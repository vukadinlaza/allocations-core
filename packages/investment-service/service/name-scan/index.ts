import { Investment } from "@allocations/core-models";
import { connectMongoose } from "@allocations/service-common";
import { getNameScanData } from "../utils/namescan";
import { updateKYCData } from "../airtable";

export const handler = async ({ Records }: any): Promise<void> => {
  await connectMongoose();
  for (const record of Records) {
    const { _id } = JSON.parse(record.Sns.Message);
    const investment = await Investment.findById(_id);

    try {
      if (!investment) {
        throw new Error(
          `Unable to find investment with id ${_id} while updating KYC Data`
        );
      }
      const nameScanData = (await getNameScanData(investment)) || {};
      await updateKYCData(investment, nameScanData);
    } catch (err) {
      console.error("Error syncing KYC data to airtable", err);
    }
  }
};
