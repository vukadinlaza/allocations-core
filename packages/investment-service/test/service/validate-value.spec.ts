import * as t from "runtypes";
import { validateValue } from "../../service/validate-value";

describe("Unit: validateValue", () => {
  const TestResponseSchema = t.Record({
    status: t.Union(t.Literal("OK"), t.Literal("FAILURE")),
    code: t.Number,
    body: t.Record({
      foo: t.Array(t.String.Or(t.Null)),
    }),
  });

  it("returns the response if it matches the schema", () => {
    const testResponse = {
      status: "OK",
      code: 201,
      body: {
        foo: ["super-cool-string"],
      },
    };

    const result = validateValue({
      schema: TestResponseSchema,
      value: testResponse,
    });

    expect(result).toMatchSnapshot();
  });

  it("returns an error and success:false if response does not match schema", () => {
    const testResponse = {
      status: "OK",
      code: 404,
      body: {
        error: "you failed",
      },
    };

    const result = validateValue({
      schema: TestResponseSchema,
      value: testResponse,
    });

    expect(result).toMatchSnapshot();
  });
});
