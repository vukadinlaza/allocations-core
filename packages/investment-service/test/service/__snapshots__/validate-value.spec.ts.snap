// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit: validateValue returns an error and success:false if response does not match schema 1`] = `
Object {
  "error": [ValidationError: Expected { status: "OK" | "FAILURE"; code: number; body: { foo: (string | null)[]; }; }, but was incompatible],
  "success": false,
}
`;

exports[`Unit: validateValue returns the response if it matches the schema 1`] = `
Object {
  "success": true,
  "value": Object {
    "body": Object {
      "foo": Array [
        "super-cool-string",
      ],
    },
    "code": 201,
    "status": "OK",
  },
}
`;
